app/controllers/base_controller.rb

```diff
@@ -0,0 +1,7 @@
+class BaseController < ApplicationController
```

Comment:
controller này để làm j em

Reply:
nó có khác j với admin controller ko, tham khảo comment bên admin controller nhé 

chị thấy ở đây chỉ cần check user phải login thôi á, chị thấy controlelr này là class chung nên chị nghĩ em chỉ cần check đã login mới vào dc hệ thống thôi

Reply:
Dạ em phân ra 2 base controller, 1 bên của admin 1 bên của user để kiểm soát quyền truy cập các trang dựa trên role ạ. Với admin thì không truy cập được các trang của user và sẽ redirect to admin_root nếu cố gắng access vào view của user. Tương tự với user. Nếu em để hàm check_role này trong application controller thì nó sẽ như thế này 
`
  def check_role
    if current_user&.admin?
      redirect_to(admin_root_path)
    else
      redirect_to(root_path)
    end
  end`
thì có nghĩa là nó sẽ kiểm tra mỗi lần user/admin access vào 1 file nào đó thì redirect lại dù có access đúng quyền hay k. Vd như admin access vào home của user thì nó sẽ redirect lại trang admin_root của admin, rồi tại trang đó nó sẽ tiếp tục kiểm tra một lần nữa rồi redirect (vì các controller kế thừa từ application controller hàm check_role)
Còn nếu em chia ra 2 file base_controller và có 2 check_role như này 
`def check_role
    redirect_to(root_path) unless current_user&.admin?
  end`
thì nếu nó user sẽ bị redirect chỉ khi truy cập sai quyền và k bị redirect lại lần 2 ạ.

Reply:
e có admin:base_controller rồi mà?
sao cần thêm base này làm gì nữa